### DISCUSSION TITLE
Is there any way to reduce audio latency for osu?

### DISCUSSION THREAD

**Discussion by okolenmi** [ppy/osu#28824] [Q&A] [answered]
After all the rebalances osu is a game about hit accuracy only (_good for "hit corrector software" users_), even if not everyone can afford top devices to play.
_The gameplay is good only for noobs, peppy cult followers and "robots" who doesn't care about music at all._ 

Even if you will ignore input latency there is huge latency on audio output system - and it's already very uncomfortable on 6k pp level. Not sure sound card can fix it. Can osu support direct output technology for such cards? Can it support ASIO drivers or something like this? 
With current state of game it's better to play without song at all - I don't have any problems with offset in this case.

- With 0 offset  - late hits while hitting to the rhythm, 
- With around -20 offset - hitting keyboard perfectly in time, but hitsounds ruining the sound... i can't hear keyboard in my headphones, so I don't want to disable them.
Unsatysfing experience, I would say. 

So, I want to discover a way to **reduce audio latency as far as I can.** I can try to buy a good sound card, but can it help me in this game?

**Comment by bdach** [comment]
> The gameplay is good only for noobs, peppy cult followers and "robots" who doesn't care about music at all

🤨 with this sort of introduction i'm not sure if my time replying further isn't wasted, however we [did want to add WASAPI support](https://github.com/ppy/osu/pull/26154) at one point, but it didn't work properly in user testing so it got pushed back indefinitely until better days.

**Reply by okolenmi** [reply]
It's very sad info, really. It would be better for osu to have verified optional plugins like this. Even if it doesn't work for everyone, there is a chance it can work for me.
Audio latency is almost the only thing I hate in this game (other thing related to players, not the game). The game was fun when I was 3k pp, but now I hate to play it, because I should to ruin game rhythm with offset to get better accuracy on a map.

**Comment by sineplusx** [comment, marked as answer]
What operating system are you using? For Windows 10 you could try [REAL](https://github.com/miniant-git/REAL) to decrease audio latency. For Linux, there are already a number of threads about latency reduction, see [here](https://github.com/ppy/osu/discussions/28510) for example.

You also don't need a "better" sound card, the things that matter are the driver and settings.

**Reply by okolenmi** [reply]
I'm using Windows 11. I'm not 100% sure, but results on beatmaps say that I'm pressing keys ~0.5ms earlier with offset -7. I have used -15 before. It's not perfect as it can be by using sound card directly, but gameplay feels better with this. Thank you.

It looks like sound latency affects on aim speed. Maybe  I'm waiting for click sound before moving to next object. Or maybe it's just because the song sounds earlier than before.

![изображение](https://github.com/ppy/osu/assets/55319814/221bd27f-54a1-4eb8-8a2c-a760a62c3136)
I have checked it in sound recorder. Earlier full delay from keyboard click to hitsound start was 40ms, now it's 32ms. Osu have latency 15ms in it's sound library, Audio latency according to REAL is 2.7ms, so there are 14ms of delay from something I don't know...

Even 7ms delay reduction is good enough. With WASAPI it can be much better.
